odoo/addons/base/models/res_users.py

```diff
@@ -643,6 +643,10 @@ def _login(cls, db, login, password):
                         raise AccessDenied()
                     user = user.with_user(user)
                     user._check_credentials(password)
+                    tz = request.httprequest.cookies.get('tz') if request else None
+                    if tz in pytz.all_timezones and (not user.tz or not user.login_date):
+                        # first login or missing tz -> set tz to browser tz
+                        user.tz = tz
```

Comment:
note: this should never crash as the user is supposed to be able to [write to their own tz field](https://github.com/odoo/odoo/blob/87180029685b127dad504f128adc26b1a2980c5b/odoo/addons/base/models/res_users.py#L220), even portal users, so no need for `sudo()` after the `user.with_user(user)` above.